trl/trainer/xpo_trainer.py

```diff
@@ -220,6 +223,25 @@ def _compute_rewards(self, model_data, ref_data, context_length):
 
         return model_scores, ref_scores
 
+    def _compute_judge(self, model_data, ref_data, context_length):
+        prompts = self.processing_class.batch_decode(
```

Comment:
Is skipping special tokens a good idea for chat templates like ChatML? Otherwise I think we'll lose tokens like `<|im_start|>` and `<|im_end|>` which may confuse the judge

Reply:
right but my feeling was that models like pair-rm take in sentences as input... I can try to add the tokens back but then we also get padding tokens etc. which would need removing as well?

Reply:
That's tricky. For some tokenizers, as @lewtun says, when you skip the special tokens, it seems that it's not a conversation:

```python
from transformers import AutoTokenizer

messages = [
    [
        {"role": "user", "content": "What color is the sky?"},
        {"role": "assistant", "content": "The sky is blue."},
    ],
    [
        {"role": "user", "content": "What is the closest planet to the sun?"},
        {"role": "assistant", "content": "It is Mercury."},
    ],
]

tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-random-Mistral-7B-Instruct-v0.3")
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer.apply_chat_template(messages, padding=True)

print(tokenizer.batch_decode(input_ids, skip_special_tokens=True))
['What color is the sky? The sky is blue.', 'What is the closest planet to the sun? It is Mercury.']
print(tokenizer.batch_decode(input_ids, skip_special_tokens=False))
['</s></s></s><s>[INST] What color is the sky?[/INST] The sky is blue.</s>', '<s>[INST] What is the closest planet to the sun?[/INST] It is Mercury.</s>']
```

Having the special tokens can hinder the judge. This is particularly obvious with the llama:

```python
tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-random-Meta-Llama-3-8B-Instruct")
tokenizer.pad_token = tokenizer.eos_token
input_ids = tokenizer.apply_chat_template(messages, padding=True)

print(tokenizer.batch_decode(input_ids, skip_special_tokens=True))
['user\n\nWhat color is the sky?assistant\n\nThe sky is blue.', 'user\n\nWhat is the closest planet to the sun?assistant\n\nIt is Mercury.']
print(tokenizer.batch_decode(input_ids, skip_special_tokens=False))
['<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat color is the sky?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nThe sky is blue.<|eot_id|><|eot_id|><|eot_id|>', '<|begin_of_text|><|start_header_id|>user<|end_header_id|>\n\nWhat is the closest planet to the sun?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\nIt is Mercury.<|eot_id|>']
```

Reply:
Yes, I would remove the padding tokens, but keep the special tokens. We have a helper method here if that's useful: https://github.com/huggingface/trl/blob/88be2c07e5024cc47e7777d23815602ae22c5f11/trl/trainer/utils.py#L1381

Reply:
We've chosen another option, which is to retrieve the raw prompt (message list) and completion and format it with a simple chat template. Consequently, for a conversational dataset, regardless of the model's chat template, the judge will receive data in the following form:

```python
prompt = "User: What color is the sky\n\n"
completion = "Assistant: It is blue.\n\n"
```

see ce3798a